src/server/service/kv.rs

```diff
@@ -53,6 +53,91 @@ pub struct Service<T: RaftStoreRouter + 'static, E: Engine> {
     thread_load: Arc<ThreadLoad>,
 }
 
+#[derive(Eq, Hash)]
+struct RegionVerId {
+    region: u64,
+    epoch: u64,
+}
+
+impl PartialEq for RegionVerId {
+    fn eq(&self, other: &Self) -> bool {
+        self.region == other.region && self.epoch == other.epoch
+    }
+}
+struct MiniBatcher {
+    commands: Vec<Command>,
+    router: HashMap<RegionVerId, usize>,
+}
+
+impl MiniBatcher {
+    pub fn new() -> Self {
+        MiniBatcher {
+            commands: vec![],
+            router: HashMap::default(),
+        }
+    }
+
+    pub fn filter(
+        &mut self,
+        request_id: u64,
+        request: &mut batch_commands_request::Request,
+    ) -> bool {
+        match &mut request.cmd {
+            Some(batch_commands_request::request::Cmd::Prewrite(req)) => {
+                let ver = RegionVerId {
+                    region: req.get_context().get_region_id(),
+                    epoch: req.get_context().get_region_epoch().get_version(),
+                };
+                let mutations = req
+                    .take_mutations()
+                    .into_iter()
+                    .map(|mut x| match x.get_op() {
+                        Op::Put => Mutation::Put((Key::from_raw(x.get_key()), x.take_value())),
+                        Op::Del => Mutation::Delete(Key::from_raw(x.get_key())),
+                        Op::Lock => Mutation::Lock(Key::from_raw(x.get_key())),
+                        Op::Insert => Mutation::Insert((Key::from_raw(x.get_key()), x.take_value())),
+                        _ => panic!("mismatch Op in prewrite mutations"),
+                    })
+                    .collect();
+                let mut options = Options::default();
+                options.lock_ttl = req.get_lock_ttl();
+                options.skip_constraint_check = req.get_skip_constraint_check();
+                options.for_update_ts = req.get_for_update_ts();
+                options.is_pessimistic_lock = req.take_is_pessimistic_lock();
+                options.txn_size = req.get_txn_size();                
+                let cmd = Command::Prewrite {
+                    ctx: req.take_context(),
+                    mutations,
+                    primary: req.take_primary_lock(),
+                    start_ts: req.get_start_version(),
+                    options,
+                };
+                match self.router.get(&ver) {
+                    Some(&idx) => {
```

Comment:
If two requests try to modify the same key, they can't combine in the same mini-batch.